rfcs/0004-categorized-attributes.md

```diff
@@ -0,0 +1,217 @@
+It is a **goal** to maintain compatibility with the old data format, including support for every existing attribute with engine support.
+
+It is a **non-goal** to provide such access to every attribute.
+It is a **non-goal** to limit the supported set of attributes.
+It is a **non-goal** to coerce the format of the string attributes to allow for generating the attribute names programmatically.
+It is a **non-goal** to maintain parity with the old data format; new attributes might receive engine support wihtout a corresponding old attribute name.
+
+# Motivation
+
+The game is rapidly expanding the number of attributes that require in-engine support. One great example can be seen in the damage types: we have shield damage, hull damage, ion damage, fuel damage, slowing damage, heat damage, etc. These damage types all have their own damage, protection, resistance etc. attributes for parity. Currently, all of these require a unique, non machine-generated string name that is used as a literal across multiple engine functions. This usage is costly to maintain due to the number of engine changes required for a new attribute, and also leaves great potential for errors due to the use of arbitrary string literals.
+
+This RFC proposes a unified system replacing the string literals used for attribute access inside the engine, offering better maintainability and, in the future, better performance as well.
+
+# Detailed Design
+
+A _categorized attribute_ is an attribute that is affected by the proposed changes, as outlined in the goals section.
+
+This RFC identifies the various components of a _categorized attribute_ as:
+- Category (when is the change made?)
+- Effect (what is changed?)
+- Modifier (how is this changed?)
+
+_Categorized attributes_ are referred to as _old-form_ when denoted by a string, and _new-form_ when denoted by a _category_+_effect_+_modifier_ combination.
+
+Every _categorized attribute_ MUST be represented by exactly one _category_+_effect_+_modifier_ combination, though some combinations SHOULD NOT be supported.
+
+There MUST be a _default modifier_ indicating the base value of a non over-time effect.
+
+## Data format
+
+The data format MUST define a keyword for every _category_. It MUST either define a keyword for every _effect_ and _modifier_ individually, or a keyword for every _effect_ and _modifier_ combination. It MUST NOT define a keyword for the _default modifier_.
+
+These keywords MUST be supported for parsing, and MUST NOT be context-sensitive. These keywords MUST be supported by a parser in a nested approach.
+
+### Parsers
+
+A parser MAY follow one of the example formats:
+
+```
+ship <name>
+	attributes
+		<category-keyword> [<value#>]
+			<effect-keyword> [<value#>]
+				<modifier-keyword> [<value#>]
+```
+
+```
+ship <name>
+	attributes
+		<category-keyword> [<value#>]
+			<effect-modifier-keyword> [<value#>]
```

Comment:
I don't see why we want to have this version, if the version above also exists.
Having the keywords fully separate feels more flexible, and every redundant option that we don't support saves everyone time.

Reply:
Let's just keep the beautiful" keywords like "ionization' and "corrosion" and leave the "over time" part out of the keywords. I feel that the beautiful keywords will work better in the datafiles.